ChopChop/Models/Ingredients/Ingredient.swift

```diff
@@ -0,0 +1,271 @@
+
+            batches.append(addedBatch)
+        }
+    }
+
+    /**
+     Subtracts some quantity from a batch, identified by its expiry date.
+     Removes the batch if all its quantity is subtracted.
+     - Throws:
+        - `QuantityError.differentQuantityTypes`:
+            if the given quantity does not have the same type as the ingredient.
+        - `QuantityError.negativeQuantity`:
+            if the result is negative.
+        - `IngredientError.nonExistentBatch`:
+            if a batch with the given expiry date does not exist.
+        - `IngredientError.insufficientQuantity`:
+            if the batch with the given expiry date does not contain enough quantity.
+     */
+    func subtract(quantity: Quantity, expiryDate: Date?) throws {
+        guard self.quantityType == quantity.type else {
+            throw QuantityError.differentTypes
+        }
+
+        guard let batch = getBatch(expiryDate: expiryDate) else {
+            throw IngredientError.nonExistentBatch
+        }
+
+        guard batch.quantity >= quantity else {
+            throw IngredientError.insufficientQuantity
+        }
+
+        try batch.subtract(quantity)
+
+        if batch.isEmpty {
+            removeBatch(expiryDate: expiryDate)
+        }
+    }
+
+    /**
+     Uses up the given quantity, subtracting from the batch with the nearest expiry date to the current date first.
+     Removes a batch if all its quantity is subtracted.
+     - Throws:
+        - `QuantityError.differentQuantityTypes`:
+            if the given quantity does not have the same type as the ingredient.
+        - `QuantityError.negativeQuantity`: if the result is negative.
+        - `IngredientError.insufficientQuantity`: if all non-expired batches combined do not contain enough quantity.
+     */
+    func use(quantity: Quantity) throws {
+        guard self.quantityType == quantity.type else {
+            throw QuantityError.differentTypes
+        }
+
+        let currentDate = Date()
+        var subtractedQuantity = quantity
+        var usedBatches: [IngredientBatch] = []
+
+        for batch in batchesByExpiryDate {
+            guard let expiryDate = batch.expiryDate else {
+                do {
+                    try subtractedQuantity -= batch.quantity
+                    usedBatches.append(batch)
+                    continue
+                } catch QuantityError.negativeQuantity {
+                    try batch.subtract(subtractedQuantity)
+                    subtractedQuantity.value = 0
+                    break
+                }
+            }
+
+            guard expiryDate > currentDate else {
+                continue
+            }
+
+            do {
+                try subtractedQuantity -= batch.quantity
+                usedBatches.append(batch)
+            } catch QuantityError.negativeQuantity {
+                try batch.subtract(subtractedQuantity)
+                subtractedQuantity.value = 0
+                break
+            }
```

Comment:
Seems like a repetition of the above logic for batches with no expiry date.

Reply:
I'll see if I can refactor it out, thanks!